### PR TITLE
clang-formt: Add a .clang-format

### PR DESCRIPTION
**Pull request by DolceTriade**

### CODE REVIEW

**Comment by slipher** [comment]
Wouldn't it make more sense to put this in Daemon so we can use the same formatting for both?

**Comment by DolceTriade** [comment]
Put in daemon, then symlink to here or duplicate?

**Comment by slipher** [comment]
Oh I was just thinking that Unvanquished can rely on the existence of Daemon but not vice-versa. But if there's some stuff where it automatically finds the config file in certain locations I don't know how that works

**Comment by DolceTriade** [comment]
For example, with this file here, we can do stuff like `git clang-format HEAD~1` or you can do `git add file1 file2 && git clang-format` which will format only the diffs.

**Comment by Viech** [comment]
How would this interfere with existing files, open branches/PRs, and `git blame`? I like autoformatting but I'm worried about history.

**Comment by DolceTriade** [comment]
@Viech That is a bigger discussion (https://github.com/Unvanquished/Unvanquished/issues/2207)

This PR is just the current q3 style that we seem to have fallen back on over the years. It doesn't affect existing code. If you touch existing code, that code will be reformatted to match the existing style.

I believe we should solve the greater problem (by discussing it in the relevant issue)

TL;DR of the issue is:
 - reformat code using agreed upon style
 - add commits to .gitignorerevs or similar file to not show up in git blame to preserve git blame/history
 - ???
 - Profit

For existing PRs, reformatting the PRs (git clang-format origin/master) should help avoid conflicts.

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  SplitEmptyNamespace: false
+BreakBeforeBraces: Custom
+ColumnLimit: 100
+PointerAlignment: Right
+SpacesInParentheses: true
+SpacesInSquareBrackets: true
+TabWidth: 4
+UseTab: AlignWithSpaces
+CommentPragmas:  '^ keep:'
+IncludeCategories:
```
Comment (slipher) [inline .clang-format:33]: It might be better to avoid include sorting for now; I bet this causes a lot of compile errors because we have non-self-contained headers.

**Review by slipher — COMMENTED** [review]
Can we disable formatting of comments? It's not very beneficial and seems to be overly aggressive with often incorrect reflows.

One difference I noticed with our current style: it wants to have 2 spaces before a `//` comment on the same line as code but we usually use 1.

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  - Regex:           '^<.*'
+    Priority:        2
+    SortPriority:    0
+    CaseSensitive:   false
+  - Regex:           '.*'
+    Priority:        3
+    SortPriority:    0
+    CaseSensitive:   false
+StatementMacros:
+  - GETTER
```
[the remark this answers is not in the document]
Comment (DolceTriade) [reply .clang-format:63]: It's for the Lua code

File: .clang-format [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  SplitEmptyFunction: false
+  SplitEmptyRecord: false
+  SplitEmptyNamespace: false
+BreakBeforeBraces: Custom
+ColumnLimit: 100
+PointerAlignment: Right
+SpacesInParentheses: true
+SpacesInSquareBrackets: true
+TabWidth: 4
+UseTab: AlignWithSpaces
```
Comment (illwieckz) [inline .clang-format:31]: > `UseTab: AlignWithSpaces`

Does that mean every tab will be replaced with spaces?

File: .clang-format (hunk 2 above)
[the remark this answers is not in the document]
Comment (slipher) [reply .clang-format:63]: Maybe you could just put a (technically unnecessary) semicolon after these in your code.

File: .clang-format (hunk 3 above)
Comment (DolceTriade) [reply .clang-format:31]: It means it will use tabs to indent, but then will use spaces to align outside these tabs. This will only look correct if your tab width is 4.

This could be our opportunity to settle tabs vs spaces.

Personally, I used to be all tabs, but now I'm down with spaces.

Comment (illwieckz) [reply .clang-format:31]: Well, “tab for indentation, space for ASCII art” is what I expect. The information the tab provides to the user is so precious to destroy it with spaces, so I would have been against a “replace tabs with space” rule. There can't be any any spacing issue to fear if tabs are only used for indentation and spaces only used for ASCII art. So if that's what this option does, that's good.

**Comment by bmorel** [comment]
> Wouldn't it make more sense to put this in Daemon so we can use the same formatting for both?

No, because daemon and unvanquished in practice use a different coding style.

**Comment by DolceTriade** [comment]
Feel free to add comments if there are aspects of the code style you want to change.

**Comment by illwieckz** [comment]
To sum up the purpose of this change, here are some selected quotes from IRC:

```
<illwieckz> Ishq, is this clang format only used for committed changes?
<Ishq> illwieckz: yes
<Ishq> only affects new code
```

This restyle stuff is currently only for properly styling new code.

```
<perturbed> it's the old quake-derived style that Ishq is proposing
```

This restyle stuff should not changes things a lot, just avoid things to differ from the usual existing style.

```
<freem> changing a mostly used codestyle only brings problems, that's my opinion
<Ishq> reformatting the codebase has implications for branches and forks
<Ishq> so we shouldn't do that lightly
<illwieckz> yes, strong restyles are fork makers
<Ishq> enforcing this style on newly committed code seems a much lower barrier of entry
<perturbed> yes, in the wiki page i wrote about this i said we should have a branch rewriting script
<perturbed> if we go for reformatting
```

There is a kind of consensus we should not do a full restyle without minding it deeply, so that's not the purpose of that PR.

**Comment by Viech** [comment]
This PR should probably add to README.md instructions to use this: the command to run or how to setup a pre-commit hook, etc.
